File: apps/api/app/schemas/document_relation.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

RelationType = Literal["references", "appendix_of", "implements", "related"]


class RelatedDocumentSummary(BaseModel):
    id: str
    title: str
    document_number: str | None = None
    document_type: str
    status: str


class DocumentRelationOutgoingRead(BaseModel):
    id: str
    relation_type: str
    notes: str | None = None
    target_document: RelatedDocumentSummary
    created_at: datetime


class DocumentRelationIncomingRead(BaseModel):
    id: str
    relation_type: str
    notes: str | None = None
    source_document: RelatedDocumentSummary
    created_at: datetime


class DocumentRelationsResponse(BaseModel):
    document_id: str
    outgoing: list[DocumentRelationOutgoingRead] = Field(default_factory=list)
    incoming: list[DocumentRelationIncomingRead] = Field(default_factory=list)


class DocumentRelationCreateRequest(BaseModel):
    target_document_id: str
    relation_type: RelationType
    notes: str | None = Field(default=None, max_length=2000)


class DocumentRelationDeleteRead(BaseModel):
    id: str
    source_document_id: str
    target_document_id: str
    relation_type: str
    notes: str | None = None
    created_at: datetime

File: apps/web/types/document-relation.ts
export type RelationType = 'references' | 'appendix_of' | 'implements' | 'related'

export interface RelatedDocumentSummary {
  id: string
  title: string
  document_number?: string | null
  document_type: string
  status: string
}

export interface DocumentRelationOutgoing {
  id: string
  relation_type: RelationType
  notes?: string | null
  target_document: RelatedDocumentSummary
  created_at: string
}

export interface DocumentRelationIncoming {
  id: string
  relation_type: RelationType
  notes?: string | null
  source_document: RelatedDocumentSummary
  created_at: string
}

export interface DocumentRelationsResponse {
  document_id: string
  outgoing: DocumentRelationOutgoing[]
  incoming: DocumentRelationIncoming[]
}

export interface DocumentRelationCreateInput {
  target_document_id: string
  relation_type: RelationType
  notes?: string
}

export interface DocumentRelationDeleteResponse {
  id: string
  source_document_id: string
  target_document_id: string
  relation_type: RelationType
  notes?: string | null
  created_at: string
}

export type ConfidenceTier = 'high' | 'review'

export interface RelationSuggestion {
  target_document_id: string
  relation_type: RelationType
  confidence: number
  confidence_tier: ConfidenceTier
  matched_reference: string
  source_chunk_id: string
  source_chunk_quote: string
  target_document_preview: RelatedDocumentSummary
  reasons: string[]
}

export interface RelationSuggestionsResponse {
  document_id: string
  suggestions: RelationSuggestion[]
  candidate_count: number
}
